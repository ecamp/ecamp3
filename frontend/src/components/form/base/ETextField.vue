<template>
  <ValidationProvider
    v-slot="{ errors: veeErrors }"
    ref="validationProvider"
    tag="div"
    :name="name"
    :vid="veeId"
    :rules="veeRules"
    :required="required"
    class="e-form-container"
  >
    <v-text-field
      ref="textField"
      v-bind="$attrs"
      :filled="filled"
      :required="required"
      :hide-details="hideDetails"
      :error-messages="veeErrors.concat(errorMessages)"
      :label="label || name"
      :class="[inputClass]"
      :type="type"
      v-on="inputListeners"
    >
      <!-- passing through all slots -->
      <slot v-for="(_, name) in $slots" :slot="name" :name="name" />
      <template v-for="(_, name) in $scopedSlots" :slot="name" slot-scope="slotData">
        <slot :name="name" v-bind="slotData" />
      </template>
    </v-text-field>
  </ValidationProvider>
</template>

<script>
import { ValidationProvider } from 'vee-validate'
import { formComponentPropsMixin } from '@/mixins/formComponentPropsMixin.js'
import { formComponentMixin } from '@/mixins/formComponentMixin.js'

export default {
  name: 'ETextField',
  components: { ValidationProvider },
  mixins: [formComponentPropsMixin, formComponentMixin],
  props: {
    type: {
      type: String,
      default: 'text',
    },
  },
  data() {
    return {
      preventValidationOnBlur: false,
    }
  },
  computed: {
    inputListeners: function () {
      const vm = this
      return Object.assign(
        {},
        // attach all $parent listeners
        this.$listeners,
        // override @input listener for correct handling of numeric values
        {
          input: function (value) {
            vm.$data.preventValidationOnBlur = false
            vm.$emit('input', value)
          },
          blur: function () {
            vm.$emit('blur')
            if (vm.$data.preventValidationOnBlur) {
              vm.$refs.validationProvider.reset()
            }
            vm.$data.preventValidationOnBlur = false
          },
        }
      )
    },
  },
  mounted() {
    this.preventValidationOnBlur =
      'autofocus' in this.$attrs &&
      'required' in this.$refs.validationProvider.$attrs &&
      this.$refs.textField.value == ''
  },
  methods: {
    focus() {
      this.$refs.textField.focus()
    },
  },
}
</script>

<style scoped>
[required]:deep(label::after) {
  content: '\a0*';
  font-size: 12px;
  color: #d32f2f;
}
[required]:deep(.v-input--is-label-active label::after) {
  color: gray;
}
</style>
