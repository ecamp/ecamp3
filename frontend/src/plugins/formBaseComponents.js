// Global registration of all Vue components in folder components/form/base

import ECheckbox from '@/components/form/base/ECheckbox.vue'
import EColorPicker from '@/components/form/base/EColorPicker.vue'
import EDatePicker from '@/components/form/base/EDatePicker.vue'
import EForm from '@/components/form/base/EForm.vue'
import ERichtext from '@/components/form/base/ERichtext.vue'
import ESelect from '@/components/form/base/ESelect.vue'
import ESwitch from '@/components/form/base/ESwitch.vue'
import ETextarea from '@/components/form/base/ETextarea.vue'
import ETextField from '@/components/form/base/ETextField.vue'
import ETimePicker from '@/components/form/base/ETimePicker.vue'

class FormBaseComponentsPlugin {
  install(Vue) {
    Vue.component('ECheckbox', ECheckbox)
    Vue.component('EColorPicker', EColorPicker)
    Vue.component('EDatePicker', EDatePicker)
    Vue.component('EForm', EForm)
    Vue.component('ERichtext', ERichtext)
    Vue.component('ESelect', ESelect)
    Vue.component('ESwitch', ESwitch)
    Vue.component('ETextarea', ETextarea)
    Vue.component('ETextField', ETextField)
    Vue.component('ETimePicker', ETimePicker)
  }
}

export default new FormBaseComponentsPlugin()
