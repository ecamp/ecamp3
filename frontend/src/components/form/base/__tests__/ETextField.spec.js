import Vue from 'vue'
import Vuetify from 'vuetify'

import formBaseComponents from '@/plugins/formBaseComponents'

import { mount as mountComponent } from '@vue/test-utils'
import ETextField from '../ETextField.vue'

Vue.use(Vuetify)
Vue.use(formBaseComponents)

describe('An ETextField', () => {
  let vuetify

  const mount = (options, number = false) => {
    const app = Vue.component('App', {
      components: { ETextField },
      data: function () {
        return {
          data: null,
        }
      },
      template: number
        ? `
        <div data-app>
        <e-text-field v-model.number="data"/>
        </div>`
        : `
        <div data-app>
        <e-text-field v-model="data"/>
        </div>`,
    })
    return mountComponent(app, { vuetify, attachTo: document.body, ...options })
  }

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  describe('text', () => {
    test('looks like a textfield', async () => {
      const wrapper = mount()
      expect(wrapper).toMatchSnapshot('empty')

      await wrapper.setData({ data: 'MyText' })
      expect(wrapper).toMatchSnapshot('with text')
    })

    test('updates text when vModel changes', async () => {
      const wrapper = mount()
      expect(wrapper.find('.e-form-container').element.getAttribute('value')).toBeNull()

      const firstText = 'myText'
      await wrapper.setData({ data: firstText })
      expect(wrapper.find('.e-form-container').element.getAttribute('value')).toBe(
        firstText
      )

      const secondText = 'myText2'
      await wrapper.setData({ data: secondText })
      expect(wrapper.find('.e-form-container').element.getAttribute('value')).toBe(
        secondText
      )
    })

    test('updates vModel when value of input field changes', async () => {
      const wrapper = mount()
      const input = wrapper.find('input')
      const text = 'bla'

      input.element.value = text
      await input.trigger('input')

      expect(wrapper.vm.data).toBe(text)
    })
  })

  describe('number', () => {
    test('looks like a textfield', async () => {
      const wrapper = mount({}, true)
      expect(wrapper).toMatchSnapshot('empty')

      await wrapper.setData({ data: 3.14 })
      expect(wrapper).toMatchSnapshot('with text')
    })

    test('updates text when vModel changes', async () => {
      const wrapper = mount({}, true)
      expect(wrapper.find('.e-form-container').element.getAttribute('value')).toBeNull()

      const firstText = 0
      await wrapper.setData({ data: firstText })
      expect(wrapper.find('.e-form-container').element.getAttribute('value')).toBe(
        `${firstText}`
      )

      const secondText = 3.14
      await wrapper.setData({ data: secondText })
      expect(wrapper.find('.e-form-container').element.getAttribute('value')).toBe(
        `${secondText}`
      )
    })

    test('updates vModel when value of input field changes', async () => {
      const wrapper = mount({}, true)
      const input = wrapper.find('input')
      const text = 3.14

      input.element.value = text
      await input.trigger('input')

      expect(wrapper.vm.data).toBe(text)
    })
  })
})
