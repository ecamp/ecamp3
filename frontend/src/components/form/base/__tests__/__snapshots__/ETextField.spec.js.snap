// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`An ETextField > number > looks like a textfield > empty 1`] = `
<div data-app>
     <div class="e-form-container">
          <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <input id
                                     type="text">
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;

exports[`An ETextField > number > looks like a textfield > with text 1`] = `
<div data-app>
     <div class="e-form-container"
          value="3.14">
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <input id
                                     type="text">
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;

exports[`An ETextField > text > looks like a textfield > empty 1`] = `
<div data-app>
     <div class="e-form-container">
          <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <input id
                                     type="text">
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;

exports[`An ETextField > text > looks like a textfield > with text 1`] = `
<div data-app>
     <div class="e-form-container"
          value="MyText">
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <input id
                                     type="text">
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;
